File: yc-core-ical/src/main/kotlin/mikufan/cx/yc/core/ical/component/AlarmMapper.kt
package mikufan.cx.yc.core.ical.component

import com.fasterxml.jackson.databind.JsonNode
import mikufan.cx.inlinelogging.KInlineLogging
import mikufan.cx.yc.core.ical.model.AlarmSetting
import mikufan.cx.yc.core.ical.model.ShiftBasedOn
import mikufan.cx.yc.core.ical.util.*
import net.fortuna.ical4j.model.component.VAlarm
import net.fortuna.ical4j.model.parameter.Related
import net.fortuna.ical4j.model.property.Action
import net.fortuna.ical4j.model.property.Description
import net.fortuna.ical4j.model.property.RelatedTo
import net.fortuna.ical4j.model.property.Trigger
import java.time.Duration

/**
 * @author CX无敌
 * 2023-01-03
 */
class AlarmMapper {
  fun createAlarm(json: YouTrackIssueJson, alarmSetting: AlarmSetting?): VAlarm? {
    if (alarmSetting == null) {
      log.info { "No alarm setting for ${json.debugName}, skipping alarm creation" }
      return null
    }

    val (durationFieldName, isNegativeDurationField, defaultDuration, shiftBasedOn) = alarmSetting
    log.info { "Try creating VAlarm based on alarm setting for ${json.debugName}" }
    val (duration, descStr) = tryGetDurationAndDescription(
      durationFieldName,
      json,
      isNegativeDurationField,
      defaultDuration,
    ) ?: return null
    val fullDescription = when (shiftBasedOn) {
      ShiftBasedOn.START -> "$descStr start date"
      ShiftBasedOn.END -> "$descStr end date"
    }
    val trigger = Trigger(duration).apply {
      if (shiftBasedOn == ShiftBasedOn.END) {
        add<RelatedTo>(Related.END)
      }
    }

    return VAlarm().apply {
      addProperty(Action(Action.VALUE_DISPLAY))
      addProperty(trigger)
      addProperty(Description(fullDescription))
    }
  }

  private fun tryGetDurationAndDescription(
    durationFieldName: String,
    json: YouTrackIssueJson,
    isNegativeDurationField: Boolean,
    defaultDuration: Duration?,
  ): Pair<Duration, String>? {
    val descriptionFirstPart = "A Reminder for ${json["idReadable"].asText()}"
    return if (durationFieldName.isNotBlank()) {
      val durationFromField = tryGetDurationFrom(json, durationFieldName, isNegativeDurationField)
      if (durationFromField != null) {
        log.info { "Creating VAlarm using duration from field $durationFieldName for ${json.debugName}" }
        val descrptionLastPart =
          "with duration $durationFromField " + if (isNegativeDurationField) "before" else "after"
        durationFromField to "$descriptionFirstPart $descrptionLastPart"
      } else if (defaultDuration != null) {
        log.info { "Creating VAlarm using fallback default duration" }
        defaultDuration to "$descriptionFirstPart that fallbacks to default duration $defaultDuration related to"
      } else {
        log.info {
          "No default duration set while missing duration value " +
              "from field $durationFieldName for ${json.debugName}, skip setting the alarm"
        }
        null
      }
    } else if (defaultDuration != null) {
      log.info { "Creating VAlarm using default duration" }
      defaultDuration to "$descriptionFirstPart that fallbacks to default duration $defaultDuration related to"
    } else {
      log.warn { "This shouldn't happen, but we traded it as no alarm" }
      null
    }
  }

  private fun tryGetDurationFrom(json: YouTrackIssueJson, fieldName: String, isNegative: Boolean): Duration? {
    val field = json.tryGetCustomField(fieldName, YouTrackType.PERIOD_ISSUE_CUSTOM_FIELD)
    if (field.isNull()) {
      log.debug { "${json.debugName} doesn't have the custom field $fieldName for creating alarm" }
      return null
    }
    val valueJson = field["value"] as JsonNode?
    if (valueJson.isNull()) {
      log.debug { "${json.debugName} have empty or null value in the custom field $fieldName for creating alarm" }
      return null
    }
    val durationString = valueJson["id"].asText()
    return Duration.parse(durationString).let {
      if (isNegative) {
        it.negated()
      } else {
        it
      }
    }
  }
}

private val log = KInlineLogging.logger()

File: src/youcal/core/ical/alarm_mapper.py
"""AlarmMapper for creating VAlarm components from YouTrack duration fields."""

import logging
from datetime import timedelta

from icalendar import Alarm

from youcal.core.api.models import YouTrackIssue
from youcal.core.ical.models import AlarmSetting, ShiftBasedOn
from youcal.core.ical.utils import (
  convert_youtrack_minutes_to_timedelta,
  extract_period_minutes,
  get_issue_debug_name,
)

logger = logging.getLogger(__name__)


class AlarmMapper:
  """Creates VAlarm components from YouTrack duration fields.

  This class handles extracting duration values from YouTrack issues
  and converting them to iCalendar VALARM components.
  """

  def create_alarm(
    self,
    issue: YouTrackIssue,
    alarm_setting: AlarmSetting | None,
  ) -> Alarm | None:
    """Create a VAlarm component from alarm settings.

    Args:
      issue: The YouTrack issue
      alarm_setting: Configuration for the alarm, or None to disable

    Returns:
      A VAlarm component, or None if alarm should be skipped
    """
    debug_name = get_issue_debug_name(issue)

    if alarm_setting is None:
      logger.info("No alarm setting for %s, skipping alarm creation", debug_name)
      return None

    logger.info("Creating VAlarm based on alarm setting for %s", debug_name)

    # Try to get duration and description
    result = self._get_duration_and_description(issue, alarm_setting)
    if result is None:
      return None

    duration, description = result

    # Create the alarm
    alarm = Alarm()
    alarm.add("action", "DISPLAY")
    alarm.add("description", description)

    # Set trigger - negative for before event, positive for after
    # The trigger is relative to DTSTART by default, or DTEND if specified
    if alarm_setting.shift_based_on == ShiftBasedOn.END:
      alarm.add("trigger", duration, parameters={"RELATED": "END"})
    else:
      alarm.add("trigger", duration)

    return alarm

  def _get_duration_and_description(
    self,
    issue: YouTrackIssue,
    alarm_setting: AlarmSetting,
  ) -> tuple[timedelta, str] | None:
    """Get the alarm duration and description text.

    Args:
      issue: The YouTrack issue
      alarm_setting: The alarm configuration

    Returns:
      Tuple of (duration, description) or None if alarm should be skipped
    """
    debug_name = get_issue_debug_name(issue)
    description_prefix = f"Reminder for {issue.id_readable}"

    # Try to get duration from field
    if alarm_setting.duration_field_name:
      duration = self._get_duration_from_field(issue, alarm_setting)
      if duration is not None:
        logger.info(
          "Creating VAlarm using duration from field '%s' for %s",
          alarm_setting.duration_field_name,
          debug_name,
        )
        direction = "before" if alarm_setting.is_negative_duration else "after"
        description = f"{description_prefix} with duration {duration} {direction}"
        return (duration, description)

      # Field specified but no value - fall back to default
      if alarm_setting.default_shift_duration is not None:
        logger.info(
          "Creating VAlarm using fallback default duration for %s", debug_name
        )
        duration = self._apply_sign(
          alarm_setting.default_shift_duration, alarm_setting.is_negative_duration
        )
        description = (
          f"{description_prefix} (fallback to default "
          f"{alarm_setting.default_shift_duration})"
        )
        return (duration, description)

      # No field value and no default
      logger.info(
        "No default duration set while missing duration value from field '%s' for %s, "
        "skipping alarm",
        alarm_setting.duration_field_name,
        debug_name,
      )
      return None

    # No field specified - use default if available
    if alarm_setting.default_shift_duration is not None:
      logger.info("Creating VAlarm using default duration for %s", debug_name)
      duration = self._apply_sign(
        alarm_setting.default_shift_duration, alarm_setting.is_negative_duration
      )
      description = f"{description_prefix} (default {alarm_setting.default_shift_duration})"
      return (duration, description)

    logger.warning("No alarm duration available for %s", debug_name)
    return None

  def _get_duration_from_field(
    self,
    issue: YouTrackIssue,
    alarm_setting: AlarmSetting,
  ) -> timedelta | None:
    """Extract and convert duration from a YouTrack period field.

    Args:
      issue: The YouTrack issue
      alarm_setting: The alarm configuration

    Returns:
      The converted timedelta, or None if field is missing/empty
    """
    debug_name = get_issue_debug_name(issue)
    field_name = alarm_setting.duration_field_name

    minutes = extract_period_minutes(issue, field_name)
    if minutes is None:
      logger.debug(
        "%s doesn't have value in field '%s' for creating alarm",
        debug_name,
        field_name,
      )
      return None

    # Convert YouTrack work minutes to real calendar duration
    duration = convert_youtrack_minutes_to_timedelta(minutes)

    # Apply sign based on configuration
    return self._apply_sign(duration, alarm_setting.is_negative_duration)

  def _apply_sign(self, duration: timedelta, is_negative: bool) -> timedelta:
    """Apply sign to duration based on configuration.

    Args:
      duration: The duration value
      is_negative: If True, return negative duration

    Returns:
      The duration with appropriate sign
    """
    if is_negative:
      return -duration
    return duration
